### PR TITLE
Metadata testcases

### PR DESCRIPTION
**Pull request by hhsecond**
## Description
Basic test cases for Metadata

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **[CONTRIBUTING](../CONTRIBUTING.rst)** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/test_metadata.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    def test_invalid_name(self, w_checkout):
+        # TODO: It's a value error in dataset and branch
+        with pytest.raises(SyntaxError):
+            w_checkout.metadata.add('a a', 'b')
+
+    def test_name_conflict(self, w_checkout):
+        w_checkout.metadata.add('a', 'b')
+        w_checkout.commit()
+        # TODO: shouldn't it raise a warning atleast?
```
Comment (rlizzo) [inline tests/test_metadata.py]: Good catch. Please create an issue and reference it to this comment!

File: tests/test_metadata.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        w_checkout.metadata.add('a', 'c')
+        w_checkout.commit('second time')
+        assert w_checkout.metadata.get('a') == 'c'
+
+    def test_other_datatype(self, w_checkout):
+        # TODO: should raise a proper error
```
Comment (rlizzo) [inline tests/test_metadata.py]: Good catch. Please create an issue and reference it to this comment!

File: tests/test_metadata.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        with pytest.raises(AttributeError):
+            w_checkout.metadata.add(1, '1')
+        with pytest.raises(AttributeError):
+            w_checkout.metadata.add('1', 1)
+
+    def test_read_only_mode(self, written_repo):
+        co = written_repo.checkout()
+        # TODO: should be permissionerror
```
Comment (rlizzo) [inline tests/test_metadata.py]: Good catch. Please create an issue and reference it to this comment!

Probably will be fixed by a solution to https://github.com/tensorwerk/hangar-py/pull/25#discussion_r281058815

File: tests/test_metadata.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            for i, k in enumerate(co.metadata):
+                assert co.metadata[k] == f'v_{i}'
+        for i, v in enumerate(co.metadata.values()):
+            assert co.metadata[f'k_{i}'] == v
+        for i in range(limit):
+            assert f'k_{i}' in co.metadata
+
+    def test_context_manager(self, repo):
+        co = repo.checkout(write=True)
+        # TODO: what are the usecases for this context manager
```
Comment (rlizzo) [inline tests/test_metadata.py]: See these two sections of the documentation:

* https://hangar-py.readthedocs.io/en/latest/tutorial.html#performance
* https://hangar-py.readthedocs.io/en/latest/design.html#things-in-life-change-your-data-shouldn-t

The context manager form provides up to an order of magnitude speedup for read/write operations (more so for writes, but the point holds for reads as well). If you still have questions, please let me know and i'll improve the docs so they are more clear on this matter. It's a really important point for users to understand!

**Comment by hhsecond** [comment]
> Let's also add a few tests which stress the amount of text which can be in a value, and the length of a metadata key name as well. I have a sinking suspicion that the default lmdb key length limit (511 bytes) will raise errors which may not propagate clearly through to user space if they are exceeded.

Will add this to stress test cases

File: tests/test_metadata.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import pytest
+
+
+class TestMetadata(object):
+
+    def test_invalid_name(self, w_checkout):
+        # TODO: It's a value error in dataset and branch
```
Comment (hhsecond) [inline tests/test_metadata.py]: @rlizzo Why do you think it's a SyntaxError? Illegal names in dataset creation and branch creation raises ValueError

Comment (rlizzo) [reply tests/test_metadata.py]: You're right this should be a value error.

**Comment by hhsecond** [comment]
@rlizzo This can be merged if you are OK with the review

**Comment by rlizzo** [comment]
Looks good. Let's merge!

**Pull request merged**